### PR TITLE
Experimental: Enable instance checkpointing with DMTCP

### LINKED ISSUE #16: Add checkpointing with DMTCP
This is a continuation of apptainer/singularity#468.

### PR DESCRIPTION
**Pull request by ikaneshiro**
### Overall approach

This PR introduces features to inject dmtcp binaries and libraries from the host into container instances upon startup to enable the processes inside to be checkpointed and restarted.

- Creates dmtcp-config.yaml to manage dmtcp binaries and libraries to inject into containers.
- Introduces the `--dmtcp-launch` and `--dmtcp-restart` flags for `instance start`.
- Creates the `checkpoint` command group for managing checkpoint state and checkpointing running instances.
- Starts dmtcp coordinator with a random system assigned port to prevent coordinator port collisions on the node when using the host network namespace.

### Near term work
- Allow configuration/flag driven dmtcp_launch options to enable plugins such as [CRAC](https://github.com/JainTwinkle/CRAC-early-development)

### Limitations

- dmtcp can only work with dynamically linked applications.
- dmtcp uses `dladdr1()` which is a GNU extension that requires `glibc` or `glibc` compatibility inside the container.

### DMTCP installation

Installation instructions for dmtcp can be found here: https://github.com/dmtcp/dmtcp/blob/master/INSTALL.md
In order to maximize the portability of the dmtcp build, I would recommend using the `--enable-static-libstdcxx` configure option.

In order for this feature to work the binaries must be in the users `PATH` and the libraries must be discoverable by the dynamic linker on the host.

### Examples

The following container (referred to as `server.sif` below) contains a simple http server that allows a variable to be read and written with `GET` and `POST` requests respectively:
```
Bootstrap: docker
From: python:3.10-buster

%post
    mkdir /app
    cat > /app/server.py <<EOF
import argparse
from http.server import BaseHTTPRequestHandler, HTTPServer

state = "0"

parser = argparse.ArgumentParser(description='Optional app description')
parser.add_argument('port', type=int, help='A required integer port argument')
args = parser.parse_args()

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.send_header('Content-type','text/plain')
        self.end_headers()

        self.wfile.write(bytes(state, "utf8"))
    def do_POST(self):
        self.send_response(200)
        self.send_header('Content-type','text/plain')
        self.end_headers()

        global state
        state = self.rfile.read(1).decode("utf8")
        self.wfile.write(bytes(state, "utf8"))


with HTTPServer(('', args.port), handler) as server:
    server.serve_forever()
EOF

%startscript
    python3 /app/server.py $@

```

This container can be built with:
```
$ apptainer build --fakeroot server.sif server.def
```

The following sequence of commands will take us through creating a checkpoint, starting an instance with dmtcp injected into it:
```
$ apptainer checkpoint create example-checkpoint
INFO:    Checkpoint "example-checkpoint" created.
$ apptainer instance start --dmtcp-launch example-checkpoint server.sif server 8888 # this last arg is the port the server will listen to.
INFO:    instance started successfully
```

Now that we have started our server, we can read its current state:
```
$ curl localhost:8888; echo
0
```

We will now update the state of the server from `0` to `1`:
```
$ curl -X POST localhost:8888 -d '1'; echo
1
$ curl localhost:8888; echo
1
```

Now that our server is in a different than default state, we can checkpoint it and stop it:
```
$ apptainer checkpoint instance server
INFO:    Using checkpoint "example-checkpoint"
$ apptainer instance stop server
INFO:    Stopping server instance of /home/ian/devel/src/apptainer/e2e/testdata/server.sif (PID=209072)
```

We can restart our server using out checkpoint with the following:
```
$ apptainer instance start --dmtcp-restart example-checkpoint server.sif restarted-server 8888
INFO:    instance started successfully
```

And we can verify the state is the same as we left it:
```
$ curl localhost:8888; echo
1
```

Now we can stop our instance and delete our checkpoint if we no longer want it anymore:
```
$ apptainer instance stop restarted-server
INFO:    Stopping restarted-server instance of /home/ian/devel/src/apptainer/e2e/testdata/server.sif (PID=247679)
$ apptainer checkpoint delete example-checkpoint
INFO:    Checkpoint "example-checkpoint" deleted.
```
### This fixes or addresses the following GitHub issues:

 - Fixes #16

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

### CODE REVIEW

**Comment by gmkurtzer** [comment]
Wahoo, this looks fantastic, great work @ikaneshiro!

If DMTCP isn't installed (or found), we should have an appropriate message to go along with the error of how to install it. Also, I'd like to see if DMTCP can be packaged and point to some binary RPMs or a repo that has it (perhaps EPEL here?).

**Comment by ikaneshiro** [comment]
Marking this feature set experimental to allow flexibility as community feedback may warrant breaking changes to improve the overall usability for this feature set as it matures.

File: mlocal/frags/build_runtime.mk [hunk 1]
```diff
@@ -140,3 +140,13 @@ $(global_keyring_INSTALL): $(global_keyring)
 
 INSTALLFILES += $(global_keyring_INSTALL)
 
+# dmtcp  config file
+dmtcp_conf := $(SOURCEDIR)/etc/dmtcp-conf.yaml
```
Comment (cclerget) [inline mlocal/frags/build_runtime.mk:144]: This file need to be added in https://github.com/apptainer/apptainer/blob/main/dist/nfpm/nfpm.yaml

Comment (ikaneshiro) [reply mlocal/frags/build_runtime.mk:144]: Rebased to incorporate nfpm changes and included in https://github.com/apptainer/apptainer/pull/109/commits/4bb73d153f5f9d75a1751f2d19ed3d48f0fedbe0

**Review by cclerget — APPROVED** [review]

**Pull request merged**